### PR TITLE
feat(vertexai): Add a new sample for automatic function calling

### PR DESCRIPTION
**Pull request by jaycee-li**
## Description

This PR added a new sample `AutomaticFunctionCalling.java` for automatic function calling usage. It also updated the library version and removed a deprecated field in `GroundingWithPublicData.java`.

## Checklist

- [x] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [x] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] **Tests** pass:   `mvn clean verify` **required**
- [x] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package vertexai.gemini;
+
+// [START generativeaionvertexai_gemini_automatic_function_calling]
+// [START aiplatform_gemini_automatic_function_calling]ß
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:20]:
```suggestion
// [START aiplatform_gemini_automatic_function_calling]
```

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+public class AutomaticFunctionCalling {
+  public static void main(String[] args) throws IOException, NoSuchMethodException {
+    // TODO(developer): Replace these variables before running the sample.
+    String projectId = "your-google-cloud-project-id";
+    String location = "us-central1";
+    String modelName = "gemini-1.5-flash-001";
+
+    String promptText = "What's the weather like in Paris?";
+
+    whatsTheWeatherLike(projectId, location, modelName, promptText);
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:43]: Make the method name generic.

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  /** Callable function getCurrentWeather. */
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:46]: Can we add a comment clarifying this is just a stub and external functions with http calls can be substituted here?

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  public static String getCurrentWeather(String location) {
+    if (location.equals("Paris")) {
+      return "raining";
+    } else {
+      return "sunny";
+    }
+  }
+
+  // A request involving the interaction with an external tool
+  public static String whatsTheWeatherLike(
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:56]: Update method name.

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  /** Callable function getCurrentWeather. */
+  public static String getCurrentWeather(String location) {
+    if (location.equals("Paris")) {
+      return "raining";
+    } else {
+      return "sunny";
+    }
+  }
+
+  // A request involving the interaction with an external tool
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:55]: Update the description to provide an overview of function calling.

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    // This client only needs to be created once, and can be reused for multiple requests.
+    try (VertexAI vertexAI = new VertexAI(projectId, location)) {
+      // Get the callable method instance
+      Method function = AutomaticFunctionCalling.class.getMethod("getCurrentWeather", String.class);
+      // Use the fromFunc helper method to create a FunctionDeclaration
+      FunctionDeclaration functionDeclaration =
+          FunctionDeclarationMaker.fromFunc(
+              "Get the current weather in a given location", function, "location");
+      System.out.println("Function declaration:");
+      System.out.println(functionDeclaration);
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:69]:
```suggestion
      System.out.printf("Function declaration: %s", functionDeclaration);
```

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      ChatSession chat = model.startChat();
+
+      System.out.println(String.format("Ask the question: %s", promptText));
+      // Send message with the responder, which auto-responds FunctionCalls and
+      // returns the final text result
+      GenerateContentResponse response =
+          chat.withAutomaticFunctionCallingResponder(responder).sendMessage(promptText);
+
+      // Check the final response
+      System.out.println("Print response: ");
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:90]:
```suggestion
```

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      String finalAnswer = ResponseHandler.getText(response);
+      System.out.println(finalAnswer);
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:92]:
```suggestion
      System.out.printf"Response: %s", finalAnswer);
```

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      // Start a chat session from a model, with the use of the declared function.
+      GenerativeModel model =
+          new GenerativeModel(modelName, vertexAI).withTools(Arrays.asList(tool));
+      ChatSession chat = model.startChat();
+
+      System.out.println(String.format("Ask the question: %s", promptText));
+      // Send message with the responder, which auto-responds FunctionCalls and
+      // returns the final text result
+      GenerateContentResponse response =
+          chat.withAutomaticFunctionCallingResponder(responder).sendMessage(promptText);
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:87]: If the `withAutomaticFunctionCallingResponder(responder)` is not chained here, will the function be auto-inferred? What's the use of chaining the function here when it's already present as part of the tool?

File: vertexai/snippets/src/main/java/vertexai/gemini/GroundingWithPublicData.java [hunk 10]
```diff
@@ -43,15 +43,16 @@ public static String groundWithPublicData(String projectId, String location, Str
     // Initialize client that will be used to send requests.
     // This client only needs to be created once, and can be reused for multiple requests.
     try (VertexAI vertexAI = new VertexAI(projectId, location)) {
-      Tool googleSearchTool = Tool.newBuilder()
-          .setGoogleSearchRetrieval(
-              // Enable using the result from this tool in detecting grounding
-              GoogleSearchRetrieval.newBuilder().setDisableAttribution(false))
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/GroundingWithPublicData.java:49]: Why are we turning off the attribution here?

**Review by Sita04 — CHANGES_REQUESTED** [review]

**Comment by Sita04** [comment]
Kindly fix the test failure.

**Comment by Sita04** [comment]
@jaycee-li gentle ping on this PR.

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java (hunk 9 above)
Comment (jaycee-li) [reply vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:88]: No, the function won't be auto-inferred if `withAutomaticFunctionCallingResponder(responder)` is not presented.

While adding the tool introduces the "Function Calling" feature, enabling the model to potentially respond with a `FunctionCall` instruction, users still need to manually execute the function, construct a `FunctionResponse`, and make another request.

Adding the responder activates "Automatic Function Calling" feature, automating these manual steps within the client library, saving users time and code. However, as not all functions are callable within the same Java runtime, this feature is disabled by default, users need to explicit chain it when making the request.

File: vertexai/snippets/src/main/java/vertexai/gemini/GroundingWithPublicData.java (hunk 10 above)
Comment (jaycee-li) [reply vertexai/snippets/src/main/java/vertexai/gemini/GroundingWithPublicData.java:49]: This is a preview attribution for testers, and the backend has removed this field.

**Comment by jaycee-li** [comment]
> @jaycee-li gentle ping on this PR.

Hi @Sita04 , apologies for the delay, I was OOO for the past 4 weeks.

Thanks for the review! I've addressed all the comments.

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java [hunk 11]
```diff
@@ -52,8 +54,8 @@ public static String getCurrentWeather(String location) {
     }
   }
 
-  // A request involving the interaction with an external tool
-  public static String whatsTheWeatherLike(
+  // Use the Automatic Function Calling feature to auto-response Function Call requests from the model.
```
Comment (Sita04) [inline vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:57]:
```suggestion
  // Use the Automatic Function Calling feature to auto-respond to Function Call requests from the model.
```

File: vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java (hunk 9 above)
Comment (Sita04) [reply vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:88]: Thanks for the clarification!

**Review by Sita04 — APPROVED** [review]
One nit. LGTM

**Comment by jaycee-li** [comment]
Hi @Sita04, the previous integration test failed and I've fixed the issue. Could you please re-run the test? Thank you!

**Pull request merged**